modules/assignment/src/main/java/com/intuit/wasabi/assignment/impl/AssignmentsImpl.java

```diff
@@ -866,11 +866,15 @@ protected Experiment getExperiment(Application.Name applicationName, Experiment.
         Experiment result = null;
 
         if (metadataCacheEnabled) {
-            List<Experiment> experiments = metadataCache.getExperimentsByAppName(applicationName);
-            for (Experiment exp : experiments) {
-                if (experimentLabel.equals(exp.getLabel())) {
-                    result = exp;
-                    break;
+            //First fetch experiment list sorted by priorities
+            Optional<PrioritizedExperimentList> prioritizedExperimentListOptional = metadataCache.getPrioritizedExperimentListMap(applicationName);
+            if (prioritizedExperimentListOptional.isPresent()) {
+                //Iterate as per experiment priority and look for the matching experiment by their label
+                for (PrioritizedExperiment prioritizedExperiment : prioritizedExperimentListOptional.get().getPrioritizedExperiments()) {
+                    if (experimentLabel.equals(prioritizedExperiment.getLabel())) {
+                        //Upon match, get the complete experiment object from cache
+                        result = metadataCache.getExperimentById(prioritizedExperiment.getID()).orElseGet(null);
```

Comment:
@bargenilesh Nit: Also, to @gittify 's point, I know this is true even for pre-existing code, but you could explicitly return an `Optional<Experiment>` from this method as it can potentially return a null value, very well serving the purpose of the `Optional` construct and making it more explicit that it can potentially return a nullable value.